module/zfs/zvol.c

```diff
@@ -1040,6 +1040,68 @@ zvol_create_snap_minor_cb(const char *dsname, void *arg)
 	return (0);
 }
 
+/*
+ * If spa_keystore_load_wkey() is called for an encrypted zvol,
+ * we need to look for any clones also using the key. This function
+ * is "best effort" - so we just skip over it if there are failures.
+ */
+static void
+zvol_add_clones(const char *dsname, list_t *minors_list)
+{
+	/* Also check if it has clones */
+	dsl_dir_t *dd = NULL;
+	dsl_pool_t *dp = NULL;
+
+	if (dsl_pool_hold(dsname, FTAG, &dp) != 0)
+		return;
+
+	if (!spa_feature_is_enabled(dp->dp_spa,
+	    SPA_FEATURE_ENCRYPTION))
+		goto out;
+
+	if (dsl_dir_hold(dp, dsname, FTAG, &dd, NULL) != 0)
+		goto out;
+
+	if (dsl_dir_phys(dd)->dd_clones == 0)
+		goto out;
+
+	zap_cursor_t *zc = kmem_alloc(sizeof (zap_cursor_t), KM_SLEEP);
+	zap_attribute_t *za = kmem_alloc(sizeof (zap_attribute_t), KM_SLEEP);
```

Comment:
Is there a reason these need to be allocated rather than being on the stack?

Reply:
Generally large structs, and kernel stack is tight.

```
../../../zfs/zvol.c:1048:1: warning: stack frame size of 344 bytes in function
      'zvol_add_clones' [-Wframe-larger-than=]
```

on stack:
```
../../../zfs/zvol.c:1048:1: warning: stack frame size of 680 bytes in function
      'zvol_add_clones' [-Wframe-larger-than=]
```